src/sleep/sleep.c

```diff
@@ -263,41 +263,83 @@ static int execute(
 }
 
 static int execute_s2h(const SleepConfig *sleep_config) {
-        _cleanup_close_ int tfd = -1;
-        struct itimerspec ts = {};
+        usec_t suspend_interval = sleep_config->hibernate_delay_sec, before_timestamp, after_timestamp;
         int r;
 
         assert(sleep_config);
 
-        tfd = timerfd_create(CLOCK_BOOTTIME_ALARM, TFD_NONBLOCK|TFD_CLOEXEC);
-        if (tfd < 0)
-                return log_error_errno(errno, "Error creating timerfd: %m");
+        do {
+                _cleanup_close_ int tfd = -1;
+                struct itimerspec ts = {};
+                int last_capacity, current_capacity, previous_discharge_rate, estimated_discharge_rate = 0;
+                double time_elapsed;
 
-        log_debug("Set timerfd wake alarm for %s",
-                  FORMAT_TIMESPAN(sleep_config->hibernate_delay_sec, USEC_PER_SEC));
+                tfd = timerfd_create(CLOCK_BOOTTIME_ALARM, TFD_NONBLOCK|TFD_CLOEXEC);
+                if (tfd < 0)
+                        return log_error_errno(errno, "Error creating timerfd: %m");
 
-        timespec_store(&ts.it_value, sleep_config->hibernate_delay_sec);
+                /* Calculate current battery capacity and current time before suspension*/
+                r = read_battery_capacity_percentage();
+                if (r < 0)
+                        return log_error_errno(r, "Error fetching battery capacity percentage: %m");
```

Comment:
ok. Yes. Will check.

Reply:
https://github.com/torvalds/linux/blob/master/drivers/acpi/battery.c#L93

These values will be defined only when one has a battery which may not be the case for desktop so adding check for -ENOENT in case file is not present. One possibility could still be challenging here if the file got corrupted / deleted.

Reply:
it's a sysfs so don't worry about corruption/deletion. The ENOENT check is still missing though?